rust/src/operations/filesystem_check.rs

```diff
@@ -0,0 +1,174 @@
+        let version = self.state.version();
+        let store = self.store.clone();
+
+        self.state.files().iter().for_each(|active| {
+            files_to_check.insert(active.path.to_owned(), active);
+        });
+
+        let mut files = self.store.list(None).await?;
+        while let Some(result) = files.next().await {
+            let file = result?;
+            files_to_check.remove(file.location.as_ref());
+        }
+
+        files_to_check
+            .into_iter()
+            .for_each(|(_, file)| files_to_remove.push(file.to_owned()));
+
+        Ok(FileSystemCheckPlan {
+            files_to_remove,
+            version,
+            store,
+        })
+    }
+}
+
+impl FileSystemCheckPlan {
+    pub async fn execute(self) -> DeltaResult<FileSystemCheckMetrics> {
+        if self.files_to_remove.is_empty() {
+            return Ok(FileSystemCheckMetrics {
+                dry_run: false,
+                files_removed: Vec::new(),
+            });
+        }
+
+        let mut actions = Vec::new();
+        let mut removed_file_paths = Vec::new();
+        let version = self.version;
+        let store = &self.store;
+
+        for file in self.files_to_remove {
+            let deletion_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
+            let deletion_time = deletion_time.as_millis() as DeltaDataTypeLong;
+            removed_file_paths.push(file.path.clone());
+            actions.push(Action::remove(Remove {
+                path: file.path,
+                deletion_timestamp: Some(deletion_time),
+                data_change: true,
+                extended_file_metadata: None,
+                partition_values: Some(file.partition_values),
+                size: Some(file.size),
+                tags: file.tags,
+            }));
+        }
+
+        if !actions.is_empty() {
+            commit(
+                store,
+                version + 1,
+                actions,
+                DeltaOperation::FileSystemCheck {},
+                None,
+            )
+            .await?;
+        }
+
+        Ok(FileSystemCheckMetrics {
+            dry_run: false,
+            files_removed: removed_file_paths,
+        })
+    }
+}
+
+impl std::future::IntoFuture for FileSystemCheckBuilder {
+    type Output = DeltaResult<(DeltaTable, FileSystemCheckMetrics)>;
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let plan = this.create_fsck_plan().await?;
+            if this.dry_run {
+                return Ok((
+                    DeltaTable::new_with_state(this.store, this.state),
+                    FileSystemCheckMetrics {
+                        files_removed: plan
+                            .files_to_remove
+                            .iter()
+                            .map(|f| f.path.clone())
```

Comment:
I think you can consume the plan with `into_iter()` and just take ownership of the plan:
```suggestion
                            .into_iter()
                            .map(|f| f.path)
```